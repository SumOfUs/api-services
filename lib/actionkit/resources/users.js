// @flow
import { AKClient as client } from './client';
import { resolveProxyShape, rejectProxyShape } from '../helpers';
import type { ProxyShape } from '../helpers';
import type { Axios, AxiosPromise } from 'axios';
import type { User, IUserUpdate, UserCollection } from '../actionkit.types.js';

export type SearchResult = Promise<ProxyShape<User[]>>;
export type SearchFilters = { [key: string]: string | number | boolean };

export function search(f: ?SearchFilters, config: any = {}) {
  return client
    .get(`/user`, {
      params: f || {},
    })
    .then(resolveProxyShape, rejectProxyShape);
}

export function find(id: string, config: any = {}) {
  return client
    .get(`/user/${id}`, {
      ...config,
    })
    .then(resolveProxyShape, rejectProxyShape);
}

export function update(id: string, data: IUserUpdate, config: any = {}) {
  return client
    .patch(`/user/${id}/`, data)
    .then(resolveProxyShape, rejectProxyShape);
}
