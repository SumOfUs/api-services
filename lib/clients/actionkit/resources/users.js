// @flow
import { AKClient as client } from './client';
import { resolveProxyShape, rejectProxyShape } from '../helpers';
import type { ProxyShape } from '../helpers';
import type { Axios, AxiosPromise } from 'axios';
import type { User, IUserUpdate, UserCollection } from '../actionkit.types.js';

export type SearchResult = Promise<ProxyShape<User[]>>;
export type SearchFilters = { [key: string]: string | number | boolean };

export function search(f: ?SearchFilters = {}) {
  return client
    .get(`/user/`, { params: f })
    .then(resolveProxyShape, rejectProxyShape);
}

export function find(id: string) {
  return client
    .get(`/user/${id}/`)
    .then(resolveProxyShape, rejectProxyShape)
    .then(response => {
      if (!response.body) {
        return Promise.reject({ ...response, statusCode: 404 });
      }
      return response;
    });
}

export function update(id: string, data: IUserUpdate, config: any = {}) {
  return client
    .patch(`/user/${id}/`, data)
    .then(resolveProxyShape, rejectProxyShape);
}

export function unsubscribe(
  email: string,
  page: ?string = process.env.UNSUBSCRIBE_PAGE_NAME
) {
  if (!page) {
    throw new Error('Unsubscribe page needs to be set.');
  }
  return client
    .post('/action/', { page, email })
    .then(resolveProxyShape, rejectProxyShape);
}
