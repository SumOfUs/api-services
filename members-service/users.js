import { pick } from 'lodash';
import { get } from 'axios-es6';
import { validateRequest } from '../shared/request-validator';
import { ok, notFound, badRequest } from '../shared/lambda-utils/responses';
import { searchUser } from './providers/actionkit/user';
import {
  LIST_MEMBERS_SCHEMA,
  UPDATE_MEMBER_SCHEMA,
  SHOW_MEMBER_SCHEMA,
} from './request-schemas';

export function index(event, context, callback) {
  validateRequest(LIST_MEMBERS_SCHEMA, event.queryStringParameters).then(
    params => {
      searchUser(pick(params, ...Object.keys(schema.properties)))
        .then(result => callback(null, result))
        .catch(failure => callback(null, { cors: true, body: failure.data }));
    },
    errors => callback(null, badRequest({ cors: true, body: errors }))
  );
}

export function show(event, context, callback) {
  validateRequest(SHOW_MEMBER_SCHEMA, event.queryStringParameters).then(
    params => {
      callback(null, {
        message: 'Go Serverless Webpack (Ecma Script) v1.0! First module!',
        event,
      });
    },
    errors => callback(null, badRequest({ cors: true, body: errors }))
  );
}

export function update(event, context, callback) {
  validateRequest(UPDATE_MEMBER_SCHEMA, event.queryStringParameters).then(
    params => {
      callback(null, {
        message: 'Go Serverless Webpack (Ecma Script) v1.0! First module!',
        event,
      });
    },
    errors => callback(null, badRequest({ cors: true, body: errors }))
  );
}
