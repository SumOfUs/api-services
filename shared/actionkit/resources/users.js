// @flow
import axios from 'axios';
import { resolveProxyShape, rejectProxyShape } from '../helpers';
import type { ProxyShape } from '../helpers';
import type { Axios, AxiosPromise } from 'axios';
import type { User, IUserUpdate, UserCollection } from '../actionkit.types.js';

const CONFIG = {
  baseUrl: process.env.AK_API_URL,
  auth: {
    username: process.env.AK_USERNAME,
    password: process.env.AK_PASSWORD,
  },
};
export type SearchResult = Promise<ProxyShape<User[]>>;
export type SearchFilters = { [key: string]: string | number | boolean };
export function search(f: ?SearchFilters, config: any = CONFIG) {
  return axios
    .get(`${config.baseUrl}/user`, {
      ...config,
      params: f || {},
    })
    .then(resolveProxyShape, rejectProxyShape);
}

export function find(id: string, config: any = CONFIG) {
  return axios
    .get(`${config.baseUrl}/user/${id}`)
    .then(resolveProxyShape, rejectProxyShape);
}

export function update(id: string, data: IUserUpdate, config: any = CONFIG) {
  return axios
    .put(`${config.baseUrl}/user/${id}`, data)
    .then(resolveProxyShape, rejectProxyShape);
}
