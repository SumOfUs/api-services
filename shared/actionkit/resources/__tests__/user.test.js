// @flow
import { search } from '../users';
import { pick } from 'lodash';

jest.mock('axios');
describe('search', () => {
  test('returns an array of matches', () => {
    const params = { email: 'example@example.com' };
    return search(params).then(result => expect(result).toMatchSnapshot());
  });

  test('returns an empty array if there are no matches', () => {
    const params = { email: 'nonexistentuser@example.com' };
    return search(params).then(result => expect(result).toMatchSnapshot());
  });

  test('passes through the error responses, rejecting the promise', () => {
    return expect(
      search({ email__startswith: true })
    ).rejects.toMatchSnapshot();
  });

  test('fails gracefully on network errors', () => {
    return expect(search()).rejects.toMatchSnapshot();
  });
});
