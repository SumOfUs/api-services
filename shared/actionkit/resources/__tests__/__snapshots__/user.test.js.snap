// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search fails gracefully on network errors 1`] = `[Error: connect ECONNREFUSED 127.0.0.1:80]`;

exports[`search passes through the error responses, rejecting the promise 1`] = `
Object {
  "body": Object {
    "error": "'startswith' is not an allowed filter on the 'email' field.",
  },
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "cache-control": "no-cache",
    "connection": "close",
    "content-type": "application/json",
    "date": "Tue, 21 Nov 2017 17:57:07 GMT",
    "server": "nginx",
    "transfer-encoding": "chunked",
    "vary": "Accept,Cookie,Accept-Encoding,User-Agent",
    "x-handled-for": "89.197.86.130",
    "x-machine-id": "ip-10-50-2-213",
  },
  "statusCode": 400,
}
`;

exports[`search returns an array of matches 1`] = `
Object {
  "body": Object {
    "meta": Object {
      "limit": 20,
      "next": null,
      "offset": 0,
      "previous": null,
      "total_count": 1,
    },
    "objects": Array [
      Object {
        "actions": "/rest/v1/action/?user=388175",
        "address1": "",
        "address2": "",
        "city": "Philadelphia",
        "country": "United States",
        "created_at": "2012-03-25T06:44:11",
        "email": "example@example.com",
        "events": "/rest/v1/event/?creator=388175",
        "eventsignups": "/rest/v1/eventsignup/?user=388175",
        "fields": Object {
          "en": "1",
        },
        "first_name": "Ashley",
        "id": 388175,
        "lang": "/rest/v1/language/100/",
        "last_name": "Mease",
        "location": "/rest/v1/location/388175/",
        "logintoken": "/rest/v1/user/388175/logintoken/",
        "middle_name": "",
        "orderrecurrings": "/rest/v1/orderrecurring/?user=388175",
        "orders": "/rest/v1/order/?user=388175",
        "phones": Array [],
        "plus4": "",
        "postal": "19145",
        "prefix": "",
        "rand_id": 692604506,
        "region": "",
        "resource_uri": "/rest/v1/user/388175/",
        "source": "starbucks 3/12",
        "state": "PA",
        "subscription_status": "bounced",
        "subscriptionhistory": "/rest/v1/subscriptionhistory/?user=388175",
        "subscriptions": "/rest/v1/subscription/?user=388175",
        "suffix": "",
        "token": ".388175.onPRSf",
        "updated_at": "2017-01-31T00:07:52",
        "usergeofields": "/rest/v1/usergeofield/?user=388175",
        "usermailings": "/rest/v1/usermailing/?user=388175",
        "useroriginal": "/rest/v1/useroriginal/388175/",
        "zip": "19145",
      },
    ],
  },
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "cache-control": "no-cache",
    "connection": "close",
    "content-type": "application/json",
    "date": "Tue, 21 Nov 2017 17:57:07 GMT",
    "server": "nginx",
    "transfer-encoding": "chunked",
    "vary": "Accept,Cookie,Accept-Encoding,User-Agent",
    "x-handled-for": "89.197.86.130",
    "x-machine-id": "ip-10-50-2-213",
  },
  "statusCode": 200,
}
`;

exports[`search returns an empty array if there are no matches 1`] = `
Object {
  "body": Object {
    "meta": Object {
      "limit": 20,
      "next": null,
      "offset": 0,
      "previous": null,
      "total_count": 0,
    },
    "objects": Array [],
  },
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "cache-control": "no-cache",
    "connection": "close",
    "content-type": "application/json",
    "date": "Tue, 21 Nov 2017 17:57:07 GMT",
    "server": "nginx",
    "transfer-encoding": "chunked",
    "vary": "Accept,Cookie,Accept-Encoding,User-Agent",
    "x-handled-for": "89.197.86.130",
    "x-machine-id": "ip-10-50-2-213",
  },
  "statusCode": 200,
}
`;
